### ISSUE TITLE
Evaluate local Ollama models for synthesis and query expansion

### ISSUE THREAD

**Issue by alecgard** [alecgard/knowledge-broker#27]
<html><head></head><body><h1>Evaluate local Ollama models for synthesis and query expansion</h1>
<p><strong>Blocked by:</strong> #26 (Move confidence scoring and metadata out of LLM synthesis)</p>
<h2>Context</h2>
<p>KB currently uses Claude Sonnet via the Anthropic API for two query-time tasks: answer synthesis and multi-query expansion. After #26 lands, the LLM's job in synthesis becomes much simpler — just prose generation with inline citations and natural-language contradiction notes. No JSON, no confidence scoring.</p>
<p>That simplified prompt is realistic for local models to handle. This issue is about systematically testing which Ollama models can replace Sonnet for both tasks, measuring both quality and latency, and establishing a default local model recommendation.</p>
<p>The target hardware is a MacBook Air M2 16GB — the kind of machine a solo developer or small-team user would run KB on. Models need to coexist in memory with <code>nomic-embed-text</code> (~275MB) and <code>qwen2.5:0.5b</code> (~400MB) for enrichment, leaving roughly 10–12GB for the synthesis model.</p>
<h2>Goals</h2>
<ol>
<li><strong>Identify the best local model</strong> for KB synthesis on 16GB Apple Silicon</li>
<li><strong>Measure quality vs Sonnet</strong> so we can document the tradeoff clearly</li>
<li><strong>Measure latency</strong> so users know what to expect</li>
<li><strong>Set a sensible default</strong> for <code>KB_LLM_PROVIDER=ollama</code> / <code>KB_OLLAMA_LLM_MODEL</code></li>
<li><strong>Test query expansion separately</strong> — it's a simpler task and may tolerate a smaller/faster model</li>
</ol>
<h2>Candidate models</h2>
<p>Based on what fits comfortably in ~10GB alongside the embedding and enrichment models:</p>

Model | Params | Size (Q4_K_M) | Notes
-- | -- | -- | --
qwen3:8b | 8B | ~5GB | Strong instruction following, good structured output
llama3.1:8b | 8B | ~4.7GB | Well-established, solid general quality
mistral:7b | 7B | ~4.1GB | Fast, but weaker on structured output
gemma3:4b | 4B | ~3GB | Smaller option — useful as a lower bound
phi-4-mini | 3.8B | ~2.3GB | Smallest viable option for comparison
qwen3:4b | 4B | ~2.5GB | If 8B models are too slow

<p>Check Ollama library for any newer models in the 4–8B range at the time of testing. The landscape moves fast.</p>
<p>For <strong>query expansion only</strong>, also test whether the enrichment model (<code>qwen2.5:0.5b</code>) can handle it — the expansion prompt is simple enough that it might work, and reusing an already-loaded model avoids cold-start latency.</p>
<h2>Evaluation approach</h2>
<h3>Phase 1: Retrieval quality (expansion models)</h3>
<p>Multi-query expansion affects retrieval. Use the existing eval framework to measure this.</p>
<p>For each candidate model:</p>
<ol>
<li>Configure <code>KB_LLM_PROVIDER=ollama</code> and <code>KB_OLLAMA_LLM_MODEL=&lt;model&gt;</code></li>
<li>Run <code>make eval</code> with the ACME corpus (63 test cases)</li>
<li>Record Hit@1, Hit@3, Hit@5, MRR, Recall@5 — overall and per-category</li>
<li>Pay particular attention to <code>vocabulary_mismatch</code> (currently Hit@1 0.82, Hit@5 1.00) — this is the category most dependent on expansion quality</li>
</ol>
<p><strong>Baseline:</strong> run the same eval with <code>KB_LLM_PROVIDER=claude</code> (Sonnet) for a clean comparison. Also run with <code>--no-expand</code> to isolate the expansion contribution.</p>
<h3>Phase 2: Synthesis quality (answer models)</h3>
<p>Synthesis quality can't be measured by the retrieval eval (which uses <code>QueryRaw</code>). This needs a separate assessment.</p>
<p><strong>Test set:</strong> Pick 10–15 representative queries across the eval categories. For each query, collect the synthesised answer from Sonnet and from each local model.</p>
<p><strong>Quality dimensions to assess:</strong></p>
<ul>
<li><strong>Correctness</strong> — does the answer accurately reflect the fragments? Any hallucinated claims?</li>
<li><strong>Citation accuracy</strong> — does it cite the right fragment IDs? Does it cite fragments it didn't actually use?</li>
<li><strong>Contradiction handling</strong> — when fragments disagree, does it flag the disagreement?</li>
<li><strong>Coherence</strong> — is the prose readable and well-structured, or garbled?</li>
<li><strong>Instruction following</strong> — does it stay within the fragments, or does it inject knowledge from its training data?</li>
</ul>
<p>Score each dimension 1–3 (1 = broken, 2 = adequate, 3 = good). Use Claude as an LLM judge for consistency across runs — prompt it with the fragments, the query, and the candidate answer, and ask it to score on those dimensions.</p>
<h3>Phase 3: Latency</h3>
<p>For each candidate model, measure:</p>
<ul>
<li><strong>Time to first token</strong> (cold start — model not loaded) and warm (model already in memory)</li>
<li><strong>Total synthesis time</strong> for a typical query (e.g. 5–10 fragments, ~2k tokens of context)</li>
<li><strong>Expansion time</strong> — how long the 3–5 rephrasing step takes</li>
<li><strong>Concurrent load</strong> — what happens when embedding + enrichment + synthesis models are all resident? Does anything get evicted?</li>
</ul>
<p>Capture timings per-query in the eval results. The eval <code>Summary</code> struct should be extended with per-query latency fields (or a separate timing report).</p>
<p>Practical latency thresholds:</p>
<ul>
<li><strong>Expansion:</strong> &lt;2s is fine (it runs before retrieval, users expect a brief pause)</li>
<li><strong>Synthesis total:</strong> &lt;10s is good, &lt;20s is acceptable, &gt;30s is likely too slow for interactive use</li>
<li><strong>Time to first token:</strong> &lt;3s feels responsive with streaming</li>
</ul>
<h3>Output format</h3>
<p>Produce a comparison table for each phase, saved as <code>eval/llm-comparison.md</code> or similar. Something like:</p>
<pre><code>## Retrieval (expansion model comparison)

| Model          | Hit@1 | Hit@3 | MRR  | vocab_mismatch Hit@1 | Expansion time (avg) |
|----------------|-------|-------|------|----------------------|---------------------|
| Sonnet         | 0.83  | 0.95  | 0.89 | 0.82                 | ~1.2s               |
| qwen3:8b       | ...   | ...   | ...  | ...                  | ...                 |
| llama3.1:8b    | ...   | ...   | ...  | ...                  | ...                 |
| (no expansion) | ...   | ...   | ...  | ...                  | N/A                 |

## Synthesis quality (LLM judge scores, 1–3)

| Model       | Correctness | Citations | Contradictions | Coherence | Instruction following | Avg |
|-------------|-------------|-----------|----------------|-----------|----------------------|-----|
| Sonnet      | ...         | ...       | ...            | ...       | ...                  | ... |
| qwen3:8b    | ...         | ...       | ...            | ...       | ...                  | ... |
</code></pre>
<h2>Implementation</h2>
<h3>Changes needed</h3>
<ul>
<li><strong><code>cmd/kb/cmd_eval.go</code></strong> — add per-query timing to eval results. Capture <code>time.Since</code> around the <code>QueryRaw</code> call (or <code>Query</code> if running synthesis eval) and store it on the <code>Result</code> struct.</li>
<li><strong><code>internal/eval/eval.go</code></strong> — add <code>QueryTimeMS int64</code> to <code>Result</code>, <code>AvgQueryTimeMS</code> and <code>P95QueryTimeMS</code> to <code>Summary</code> and <code>CategoryBreakdown</code>.</li>
<li><strong>Synthesis eval script/command</strong> — either extend <code>kb eval</code> with a <code>--synthesis</code> flag that runs <code>Query</code> instead of <code>QueryRaw</code> and captures the answer text, or write a standalone script that runs queries against different providers and collects outputs for judging. A script is probably simpler for a one-off comparison.</li>
<li><strong>LLM judge prompt</strong> — a prompt that takes (fragments, query, answer) and returns scores on the five quality dimensions. Can be a standalone script that calls the Anthropic API.</li>
</ul>
<h3>Test procedure</h3>
<pre><code class="language-bash"># Baseline: Sonnet with expansion
KB_LLM_PROVIDER=claude make eval

# Baseline: no expansion
kb eval --no-expand --no-save  

# Each candidate model
for model in qwen3:8b llama3.1:8b mistral:7b gemma3:4b phi-4-mini qwen3:4b; do
  ollama pull $model
  KB_LLM_PROVIDER=ollama KB_OLLAMA_LLM_MODEL=$model kb eval --ingest --json &gt; eval/results-$model.json
done
</code></pre>
<p>For synthesis quality, run the 10–15 selected queries through each model and save the outputs for LLM judge scoring.</p>
<h2>Decision criteria</h2>
<p>The goal is to find a model that is "good enough" — not Sonnet-equivalent. Specifically:</p>
<ul>
<li><strong>Retrieval regression:</strong> expansion with the local model should not drop Hit@1 by more than ~5% overall, and <code>vocabulary_mismatch</code> Hit@5 should stay at 1.00 (or very close)</li>
<li><strong>Synthesis quality:</strong> average LLM judge score of 2.0+ across all dimensions (i.e. "adequate" on everything, not broken on any dimension)</li>
<li><strong>Latency:</strong> total synthesis under 20s for a typical query on M2 Air 16GB</li>
<li><strong>Memory:</strong> model fits alongside <code>nomic-embed-text</code> + <code>qwen2.5:0.5b</code> without Ollama evicting models between calls</li>
</ul>
<p>If no single model meets all criteria, document the tradeoffs and consider recommending different models for expansion vs synthesis (expansion can tolerate a smaller model).</p>
<h2>Outcome</h2>
<ul>
<li>A recommended default for <code>KB_OLLAMA_LLM_MODEL</code> (currently empty)</li>
<li>Comment on this issue documenting the comparison results, hardware requirements, and model selection guidance</li>
</ul></body></html>
